### PR TITLE
BTAT-6760-new: Amended redirect url logic

### PR DESCRIPTION
**Pull request by davidmthornton**
APP CONFIG PR's

https://github.com/hmrc/app-config-production/pull/4547
https://github.com/hmrc/app-config-qa/pull/4786
https://github.com/hmrc/app-config-staging/pull/3251

### CODE REVIEW

File: app/controllers/agent/ConfirmClientVrnController.scala [hunk 1]
```diff
@@ -68,16 +68,33 @@ class ConfirmClientVrnController @Inject()(val messagesApi: MessagesApi,
...
+       redirectUrl match {
+
+        case Some(changeUrl) if changeUrl.contains("/vat-through-software/account")  =>
           user.session.get(SessionKeys.preference) match {
-            case Some(_) => Redirect(redirectUrl).removingFromSession(SessionKeys.redirectUrl)
+            case Some("yes") if hasVerifiedAgentEmail =>
+              Redirect(manageVatUrl).removingFromSession(SessionKeys.redirectUrl)
+            case Some("yes") if !hasVerifiedAgentEmail =>
+              Redirect(controllers.agent.routes.ConfirmEmailController.isEmailVerified())
+            case Some("no") =>
```
Comment (lewishay) [inline app/controllers/agent/ConfirmClientVrnController.scala]: Could you add some tests for each of these `Some` cases? Scoverage is saying they aren't tested at the moment.

File: app/controllers/agent/ConfirmClientVrnController.scala [hunk 2]
```diff
@@ -68,16 +68,33 @@ class ConfirmClientVrnController @Inject()(val messagesApi: MessagesApi,
...
+      val redirectUrl = user.session.get(SessionKeys.redirectUrl)
+      val hasVerifiedAgentEmail: Boolean = user.session.get(SessionKeys.verifiedAgentEmail).isDefined
+      val manageVatUrl = appConfig.manageVatCustomerDetailsUrl
+
+       redirectUrl match {
+
+        case Some(changeUrl) if changeUrl.contains("/vat-through-software/account")  =>
           user.session.get(SessionKeys.preference) match {
-            case Some(_) => Redirect(redirectUrl).removingFromSession(SessionKeys.redirectUrl)
+            case Some("yes") if hasVerifiedAgentEmail =>
```
[the remark this answers is not in the document]
Comment (davidmthornton) [reply app/controllers/agent/ConfirmClientVrnController.scala]: Great suggestion Lewis 💃

File: app/controllers/predicates/PreferencePredicate.scala [hunk 3]
```diff
@@ -39,14 +39,12 @@ class PreferencePredicate @Inject()(implicit val appConfig: AppConfig,
     val preference = agent.session.get(SessionKeys.preference)
     val hasVerifiedEmail = agent.session.get(SessionKeys.verifiedAgentEmail).isDefined
 
-    val featureSwitchUrl = redirectUrl
-
     if(appConfig.features.preferenceJourneyEnabled()) {
       preference match {
         case Some("no") =>
-          Future.successful(Left(Redirect(featureSwitchUrl)))
+          Future.successful(Left(Redirect(redirectUrl)))
```
[the remark this answers is not in the document]
Comment (davidmthornton) [reply app/controllers/predicates/PreferencePredicate.scala]: This broke tests for some reason

File: app/controllers/BaseController.scala [hunk 4]
```diff
@@ -19,25 +19,30 @@ package controllers
...
+        }
       } else {
-        Logger.warn("[JourneySetupController][journeySetup] redirectUrl was empty or an invalid absolute url")
+        Logger.warn("[BaseController][extractRedirectUrl] couldn't create ContinueUrl from empty string.")
         None
       }
     } catch {
-      case e: Exception =>
-        Logger.warn("[JourneySetupController][journeySetup] couldn't create ContinueUrl from what was provided.", e)
+      case _: Exception =>
```
Comment (lewishay) [inline app/controllers/BaseController.scala]: I'd say this warn logging was valid. We get an exception but could still benefit from seeing a specific warn log for it in the app so it is grouped with our other app warnings.

File: app/controllers/BaseController.scala [hunk 5]
```diff
@@ -19,25 +19,30 @@ package controllers
...
+        RedirectUrl(url).getEither(OnlyRelative | AbsoluteWithHostnameFromWhitelist(appConfig.environmentHost)) match {
+          case Right(value) =>
+            Some(value.toString())
+          case Left(_) =>
+            Logger.warn("[BaseController][extractRedirectUrl] redirectUrl was an invalid absolute url")
+            None
+        }
       } else {
-        Logger.warn("[JourneySetupController][journeySetup] redirectUrl was empty or an invalid absolute url")
+        Logger.warn("[BaseController][extractRedirectUrl] couldn't create ContinueUrl from empty string.")
```
Comment (lewishay) [inline app/controllers/BaseController.scala]: Same comment as before - do we want to log this at warn level? It's a valid scenario and we see many of these each day.

**Pull request merged**